pkg/utils/sort.go

```diff
@@ -16,28 +16,61 @@
 
 package utils
 
-import corev1 "k8s.io/api/core/v1"
+import (
+	"strings"
+
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+)
 
 func swap[T any](a, b *T) {
 	t := *a
 	*a = *b
 	*b = t
 }
 
-type EnvVarSlice []corev1.EnvVar
+type EnvVarIdxPair struct {
+	EnvVar corev1.EnvVar
+	Idx    int
+}
+
+type EnvVarSlice []EnvVarIdxPair
 
 func (s EnvVarSlice) Len() int {
 	return len(s)
 }
 
 func (s EnvVarSlice) Less(i, j int) bool {
-	return s[i].Name < s[j].Name
+	// If a depends on b or b depends on a, then use their input order
+	if s[i].DependsOn(s[j]) || s[j].DependsOn(s[i]) {
+		return s[i].Idx < s[j].Idx
+	}
```

Comment:
🤔 I don't think this is a stable comparator. What happens when there are transitive dependencies? For example, the original order is C, B, A, and A depends on B, and B depends on C. In this case, comparing A and C will give a result of A < C, but we all know that is incorrect.

Reply:
There are two types of DependsOn. The first is that A depends on B and B is previously defined. In this case, it implies B < A. The other is that A depends on B and B is not defined. In [K8S](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#envvar-v1-core), "If a variable cannot be resolved, the reference in the input string will be unchanged." Therefore, by putting B (if it exists) before A, the variable will be resolved, thus changing the value as well. We have to make sure B > A. In other words, as long as two variables have DependsOn relationships, their order is determined by the input order. A depends on B, and B depends on C doesn't necessarily imply A > B > C. I added additional test cases for better explanation.

Reply:
But I think in my case, the original order means A > B > C, and the comparator will return a wrong result 🤔 
Can you try to add such a test case to see if it always works?

```go
container: &corev1.Container{
          Env: []corev1.EnvVar{
	          {
		          Name:  "ENV_C",
		          Value: "value",
	          },
	          {
		          Name:  "ENV_B",
		          Value: "$(ENV_C)",
	          },
	          {
		          Name:  "ENV_A",
		          Value: "$(ENV_B)",
	          },
          },
},
```